File: 我的大学/algorithm2/05-动态规划/03-最长公共子序列.cpp
#include <cstdio>
#include <algorithm>
using namespace std;
const int maxn = 1e3 + 5;
int n, m;
char a[maxn], b[maxn];
int dp[maxn][maxn];

int main() {
    scanf("%d %d", &n, &m);
    scanf("%s %s", a + 1, b + 1);
    for (int i = 1; i <= n; ++ i)
        for (int j = 1; j <= m; ++ j) {
            dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
            if (a[i] == b[j])
                dp[i][j] = max(dp[i][j], dp[i - 1][j - 1] + 1);
        }
    printf("%d\n", dp[n][m]);
    return 0;
}

File: 我的大学/algorithm2/05-动态规划/03-最长公共子序列.go
package main

import "fmt"

const maxn = 1000   // 规定字符串最长不超过 1000
var dp [maxn][maxn]int

func max(a, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

func main() {
    var a, b string
    fmt.Scanf("%s", &a)
    fmt.Scanf("%s", &b)

    a = " " + a
    b = " " + b

    n, m := len(a), len(b)

    for i := 1; i < n; i++ {
        for j := 1; j < m; j++ {
            dp[i][j] = max(dp[i-1][j], dp[i][j-1])
            if a[i] == b[j] {
                dp[i][j] = max(dp[i][j], dp[i-1][j-1] + 1)
            }
        }
    }

    fmt.Println(dp[n-1][m-1])
}
